api/http/handler/registry.go

```diff
@@ -318,3 +353,243 @@ func (handler *RegistryHandler) handleDeleteRegistry(w http.ResponseWriter, r *h
 		return
 	}
 }
+
+// proxyRequestsToRegistryAPI proxy to registry and set auth header if necessary
+func (handler *RegistryHandler) proxyRequestsToRegistryAPI(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	registryID, err := strconv.Atoi(id)
+	if err != nil {
+		httperror.WriteErrorResponse(w, err, http.StatusBadRequest, handler.Logger)
+		return
+	}
+
+	registry, err := handler.RegistryService.Registry(portainer.RegistryID(registryID))
+	if err == portainer.ErrRegistryNotFound {
+		httperror.WriteErrorResponse(w, err, http.StatusNotFound, handler.Logger)
+		return
+	} else if err != nil {
+		httperror.WriteErrorResponse(w, err, http.StatusInternalServerError, handler.Logger)
+		return
+	}
+
+	client := &http.Client{
+		Timeout: registryTimeout,
+	}
+	if !registry.TLSVerification {
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+
+	// Build first request (without auth)
+	uri := strings.Replace(r.RequestURI, "/api/registries/"+id+"/", "", 1)
+	req := r
+	req.RequestURI = ""
+	u, err := url.Parse(registry.Protocol + "://" + registry.URL + "/" + uri)
+	if err != nil {
+		httperror.WriteErrorResponse(w, err, http.StatusBadRequest, handler.Logger)
+	}
+
+	req.URL = u
+	req.Host = registry.URL
+	resp, err := client.Do(req)
+	if err != nil {
+		httperror.WriteErrorResponse(w, err, http.StatusBadRequest, handler.Logger)
+	}
+
+	// Get Www-Authenticate header to find requested auth method
+	// Bearer: Need to auth to a 3rd party service, and get token, then recall original uri with Authorization header
+	// Basic: Need to recall same uri with Basic Authorization header
+	// Else: Just send response
+	authHeader := resp.Header.Get("Www-Authenticate")
+	authType := strings.Split(authHeader, " ")[0]
+	switch authType {
+	case "Bearer":
+		// Call auth URI to get token, and recall registry with token
+		parts := strings.Split(strings.Replace(authHeader, "Bearer ", "", 1), ",")
+		m := map[string]string{}
+		for _, part := range parts {
+			if splits := strings.Split(part, "="); len(splits) == 2 {
+				m[splits[0]] = strings.Replace(splits[1], "\"", "", 2)
+			}
+		}
+
+		authURL := m["realm"]
+		if v, ok := m["service"]; ok {
+			authURL += "?service=" + v
+			if v, ok = m["scope"]; ok {
+				authURL += "&scope=" + v
+			}
+		}
+
+		authReq, err := http.NewRequest("GET", authURL, nil)
+		if err != nil {
+			httperror.WriteErrorResponse(w, err, http.StatusBadRequest, handler.Logger)
+		}
+		authReq.SetBasicAuth(registry.Username, registry.Password)
+
+		resp, err = client.Do(authReq)
+		if err != nil {
+			httperror.WriteErrorResponse(w, err, http.StatusBadRequest, handler.Logger)
+		}
+		if resp.StatusCode == http.StatusOK {
```

Comment:
The auth header will not be added to the query, so this will return probably a 4xx error
To test that we have to add registry with external auth (with good credentials)
And then change password of the user

Reply:
Woh just tested, we have a 401, so portainer think a forbidden error and disconnect user
We have to manage that
We have to catch that and send a 5xx error